### PR TITLE
[api-extractor] Classify arrow functions as function kind in the doc model exports.

### PR DESCRIPTION
**Pull request by bartvandenende-wm**
## Summary

Classify arrow functions as function kind in the doc model exports.

Related issue ticket https://github.com/microsoft/rushstack/issues/1629

Screenshot 2024-02-06 at 16 54 40

## Details

The current behaviour of api-extractor is to classify arrow functions as `variable` kind. Even though syntactically correct, semantically (for documentation & usage purposes) this doesn't make any sense - and worse - it invalidates any documentation effort in the docstrings related to parameter descriptions, return types etc. 

This PR modifies the doc-model generation to identify arrow functions in the variable declarations and processes them as function kinds instead.

Example code
```typescript
/**
 * Demo log function
 * @param message - log message
 * @public
 */
export const log = (message: string): void => {console.log(message)}
```

**doc model exports**
_Previous behaviour_
```json
{
  "kind": "Variable",
  "canonicalReference": "sample-lib!log:var",
  "docComment": "/**\n * Demo log function\n *\n * @param message - log message\n */\n",
  "excerptTokens": [
    {
      "kind": "Content",
      "text": "log: "
    },
    {
      "kind": "Content",
      "text": "(message: string) => void"
    }
  ],
  "fileUrlPath": "src/index.ts",
  "isReadonly": true,
  "releaseTag": "Public",
  "name": "log",
  "variableTypeTokenRange": {
    "startIndex": 1,
    "endIndex": 2
  }
}
```

_New behaviour_
```json
  {
    "kind": "Function",
    "canonicalReference": "sample-lib!log:function(1)",
    "docComment": "/**\n * Demo log function\n *\n * @param message - log message\n */\n",
    "excerptTokens": [
      {
        "kind": "Content",
        "text": "log: (message: "
      },
      {
        "kind": "Content",
        "text": "string"
      },
      {
        "kind": "Content",
        "text": ") => "
      },
      {
        "kind": "Content",
        "text": "void"
      }
    ],
    "fileUrlPath": "src/index.ts",
    "returnTypeTokenRange": {
      "startIndex": 3,
      "endIndex": 4
    },
    "releaseTag": "Public",
    "overloadIndex": 1,
    "parameters": [
      {
        "parameterName": "message",
        "parameterTypeTokenRange": {
          "startIndex": 1,
          "endIndex": 2
        },
        "isOptional": false
      }
    ],
    "name": "log"
  }
``` 

## How it was tested

manual testing

## Impacted documentation

N/A

### CODE REVIEW

File: apps/api-extractor/src/generators/ApiModelGenerator.ts [hunk 1]
```diff
@@ -250,14 +250,28 @@ export class ApiModelGenerator {
...
+          this._processApiVariable(astDeclaration, context);
+        }
         break;
 
       default:
       // ignore unknown types
     }
   }
 
+  private _hasFunctionDeclaration(astDeclaration: AstDeclaration): ts.FunctionDeclaration | undefined {
```
Comment (iclanton) [inline apps/api-extractor/src/generators/ApiModelGenerator.ts:268]:
```suggestion
  private _tryFindFunctionDeclaration(astDeclaration: AstDeclaration): ts.FunctionDeclaration | undefined {
```

File: build-tests/api-extractor-scenarios/etc/spanSorting/api-extractor-scenarios.api.json [hunk 2]
```diff
@@ -375,27 +375,45 @@
...
+              "text": "{\n    a: number;\n    b(): string;\n}"
+            },
+            {
+              "kind": "Content",
+              "text": ") => "
+            },
+            {
+              "kind": "Content",
+              "text": "void"
             }
```
Comment (iclanton) [inline build-tests/api-extractor-scenarios/etc/spanSorting/api-extractor-scenarios.api.json:397]: Why did this get split up in a different way?

File: common/changes/@microsoft/api-extractor/bartvandenende-wm-api-extractor-arrow-func_2024-02-06-14-12.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/api-extractor",
+      "comment": "\"Classify arrow functions as 'function' kind in the doc model export\"",
```
Comment (iclanton) [inline common/changes/@microsoft/api-extractor/bartvandenende-wm-api-extractor-arrow-func_2024-02-06-14-12.json:5]:
```suggestion
      "comment": "Classify arrow functions as `function` kind in the doc model export.",
```

**Review by iclanton — APPROVED** [review]

File: build-tests/api-extractor-scenarios/etc/spanSorting/api-extractor-scenarios.api.json (hunk 2 above)
Comment (bartvandenende-wm) [reply build-tests/api-extractor-scenarios/etc/spanSorting/api-extractor-scenarios.api.json:397]: The `excerptTokens` are processed differently for `function` kinds to be able to map the tokens back to the `parameterTypeTokenRange` and `returnTypeTokenRange`

**Pull request merged**